protocols/kad/src/behaviour.rs

```diff
@@ -2419,7 +2427,34 @@ where
 
     fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
         self.listen_addresses.on_swarm_event(&event);
-        self.external_addresses.on_swarm_event(&event);
+        let external_addresses_changed = self.external_addresses.on_swarm_event(&event);
+
+        if external_addresses_changed && !self.connections.is_empty() {
+            let num_connections = self.connections.len();
+
+            log::debug!(
+                "External addresses changed, re-configuring {} connection{}",
+                num_connections,
+                if num_connections > 1 { "s" } else { "" }
+            );
+
+            self.queued_events
+                .extend(
+                    self.connections
+                        .iter()
+                        .map(|(conn_id, peer_id)| ToSwarm::NotifyHandler {
+                            peer_id: *peer_id,
+                            handler: NotifyHandler::One(*conn_id),
+                            event: KademliaHandlerIn::ReconfigureMode {
+                                external_addresses: self
+                                    .external_addresses
+                                    .iter()
+                                    .cloned()
```

Comment:
It would be nice to avoid this. We could do that by changing the internals of `ExternalAddresses` to backed by a `Vec` instead of `HashSet`. Then we could expose slices in addition to `Iterator` at the expense of a slightly more complicated implementation where we need to avoid duplicates ourselves. Do you think it is worth it?